### DISCUSSION TITLE
`Persisted` struct

### DISCUSSION THREAD

**Discussion by danieleades** [get-eventually/eventually-rs#181] [Q&A]
sorry, just a couple of small questions about the `Persisted` struct.

what's the different between the 'version' and the 'sequence number'? what are these each used for?

Also, I was looking at the `inmemory::EventStore` and how you would implement #49 on it.
I think if you change the 'backend' from

```rust
Arc<RwLock<HashMap<Id, Vec<Persisted<Id, Event>>>>>
```

to 

```rust
type Backend<Id, Event> = HashMap<Id, BTreeSet<Persisted<Id, Event>>>;
Arc<RwLock<Backend>>;
```

then the `RangeBounds` method are supported pretty naturally. (you would need to implement `Ord` for `Persisted`, or use `BTreeMap` instead).

It's *slightly* irksome that the `Id` appears twice in this type signature. Is it possible to move the `Id` out of the `Persisted` struct? so that you have

```rust
type Backend<Id, Event> = HashMap<Id, BTreeSet<Persisted<Event>>>;
Arc<RwLock<Backend>>
```

**Comment by iitsdani** [comment]
> what's the different between the 'version' and the 'sequence number'? what are these each used for?

`version` is the "version" of an Event Stream. Used mainly for optimistic concurrency and avoid data races. In other words, a "version" is the Event Stream-local monotonically increasing number of events in the stream.

`sequence_number` is the global offset of each event. An Event Store is basically an ordered list of Domain Events. The `sequence_number` represent the order of these events in the whole list.

As for `inmemory::EventStore`, and the `EventStore` implementation in general, we need to rework that one. It's using an outdated design, more on that on an upcoming issue I'll be creating.

As for the type definition:
1. It's not a big deal that the `inmemory::EventStore` is not efficient, as it should only be used for testing and not production-ready apps,
2. The need for implementing `std::ops::RangeBounds` is no more :v:
